### DISCUSSION TITLE
Specification of Javascript API for ConfigKey/BrowserExamKey checks

### DISCUSSION THREAD

**Discussion by jeroenhabets** [SafeExamBrowser/seb-win-refactoring#98] [Q&A] [answered]
Hi @danschlet,

Based on SEB's roadmap and [your comment](https://github.com/SafeExamBrowser/seb-win-refactoring/issues/84#issuecomment-842719683) on #84, we as an Exam Platform provider have decided to go ahead and plan to implement the Javascript API for ConfigKey/BrowserExamKey checks.

i) Is it safe to assume that, at least the key check part of, the Javascript API for SEB is now stable and will remain the same across iOS/MacOS/Windows? (please confirm)

ii) AFAICT this is not published on your website yet. Nor information that helps with implementing it (see also following subquestions).

ii) https://safeexambrowser.org/exams/bek_ck.html serves as a reference for Exam/LMS providers? (now and in the future)

iii) Do you know if there is a open source Exam/LMS key check implementation already somewhere?

iv) Am I correct that this JS key check in Mac/iOS is tracked by SEBMAC-323 on branch 3.0_unified? <= It would be helpful to point people to the SEB code that implements the SEB part of the Javascript key check.

**Comment by danschlet** [comment, marked as answer]
I'm happy to hear that exam platform providers are ready to go ahead and implement the Javascript API for ConfigKey/BrowserExamKey checks. We have had the same discussion with a couple of exam platform providers and we still have to start a detail discussion about the implementation for Moodle and other popular systems used together with SEB. Just that you understand: we're a rather small team and have to work on several SEB versions and SEB-related developments at the same time, so it's not easy to keep up with discussing all aspects with all the stake holders and interested providers, the same applies to documentation.

i) Yes, as long as it proves to work for real-world implementations in exam-platforms. As we are mostly non-web application developers, we need feedback from web application developers to make sure what we specify for the native SEB applications really works for you guys. I got some positive feedback, but as mentioned I'm still waiting for some discussions and implementations to happen on the side of our big stake holders (which are working with Moodle, OpenOLAT, ANS and other, mostly proprietary platforms). But as this implementation seems to work at least in test cases, I would like to stick with the specification and only expand it if necessary. But you can definitely consider it stable, it is already contained in the final release of SEB 3.0 for iOS and I'm currently merging the code to the macOS version. Regarding the Windows version we will discuss the timeline in the next days, but you will probably need to support the old header request method as well for a transitioning period (which could be several months, depending in which Windows version we can implement the API). I expect that we will be able to implement it to work exactly the same way, but cannot confirm this until it has been investigated by our Windows developer. Sometimes platform or browser engine restrictions force us to do slight changes per platform (but we try to avoid that whenever we can).

ii) It is currently only published in the [release notes](https://safeexambrowser.org/ios/ios_release_notes_en.html) of the 3.0 iOS version. As mentioned we have limited resources and I have to asap release the Mac 3.0 version. I will update the developer documentation as soon as possible.

ii) Yes, that simple html/JS page I created as a test and reference. As mentioned we are no web application developers, so it's up to the community to maybe provide more server-side implementation examples. 

iii) As far as I know there's no open source implementation yet, I only know of two proprietary exam software providers which apparently already implemented it.

iv) Yes, the issue is SEBMAC-323, on branch 3.0_iOS is the final version from SEB 3.0 for iOS and on 3.0_unified I'm working on the merge with the macOS code (but the macOS version is work in progress). The code can be found in [this file](https://github.com/SafeExamBrowser/seb-mac/blob/d1c90898fafc431d1e2c1a9d259e9302d93d5631/Classes/BrowserComponents/SEBAbstractModernWebView.swift#L16), I will likely place the API in a separate module later when I start expanding it.
